src/bootstrap/setup.rs

```diff
@@ -71,13 +82,34 @@ impl FromStr for Profile {
 
 impl fmt::Display for Profile {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Profile::Compiler => write!(f, "compiler"),
-            Profile::Codegen => write!(f, "codegen"),
-            Profile::Library => write!(f, "library"),
-            Profile::User => write!(f, "user"),
-            Profile::Tools => write!(f, "tools"),
+        f.write_str(self.as_str())
+    }
+}
+
+impl Step for Profile {
+    type Output = ();
+    const DEFAULT: bool = true;
+
+    fn should_run(mut run: ShouldRun<'_>) -> ShouldRun<'_> {
+        for choice in Profile::all() {
+            run = run.alias(choice.as_str());
         }
+        run
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        if run.builder.config.dry_run {
+            return;
+        }
+        let profile: Profile = match run.path.into_os_string().into_string().unwrap().parse() {
+            Ok(profile) => profile,
+            Err(_) => t!(interactive_path()),
```

Comment:
I just realized why this works - when `StepDescription::maybe_run` gets a `PathSet::empty()`, it will call `path()` on it which returns `builder.config.src`, which fails to parse. Then we try and parse it here, which fails because it's not a known profile. That seems ... very unintuitive. We should either
a) make it more clear that this is a hack by checking for `builder.config.src` explicitly and adding a comment, or
b) wait for https://github.com/rust-lang/rust/pull/96501 so RunConfig has a `Vec<PathBuf>` instead of a single path, and we can pass in 0 paths.

My preference is for b.

Reply:
I agree, I will update the code once #95601 is done